src/lex.cpp

```diff
@@ -24,35 +41,84 @@ std::vector<token> lex() {
+            continue;
+        }
+
+        if (!isspace(i)) {
+
+            //if valid operator/parenthesis input, add to output queue:
+            if (valid.find(i) != valid.end()) {
+                lexTokens.push(token(line,column, std::string(1, i)));
             }
 
-            else if(isdigit(i)){
-                placeholder = "/0";
-                bool decimal = false;
-                placeholder += i;
-                while (std::cin >> i && (isdigit(i) || i == '.')){
-                    if (i == '.') {
-                        if (decimal){
-                            //error
+            //if number input or decimal input, check validity:
+            else if (isdigit(i) || i == '.') {
+                bool decimal = (i == '.'); // sets boolean equal to true if i == '.'
+                std::string placeholder(1, i); // this converts a char (1) to a string
+                int startingColumn = column;
+
+                // checking for leading decimal
+                if (i == '.' && !(isdigit(placeholder.back()))) {
+                    std::cout << "Syntax error on line " << line << " column " << column << ".\n";
+                    exit(1);
+                }
+
+                // if number is a decimal:
+                while ((isdigit(std::cin.peek()) || std::cin.peek() == '.')) {
+                    // peek returns next character in the input sequence, without extracting it
+                    // https://cplusplus.com/reference/istream/istream/peek/
+
+                    if (std::cin.peek() == '.') {
+                        // checking for multiple decimals
+                        if (decimal) {
+                            std::cout << "Syntax error on line " << line << " column " << column + 1 << ".\n";
+                            exit(1);
                         }
                         else {
                             decimal = true;
                         }
                     }
-                    placeholder += i;
+                    char nextChar;
+                    std::cin.get(nextChar);
+                    placeholder += nextChar;
+                    column++;
                 }
-                output.push_back(maketoken(placeholder,line,column));
-            }
 
+                // checking for trailing decimal
+                if (placeholder.back() == '.') {
+                    std::cout << "Syntax error on line " << line << " column " << column + 1 << ".\n";
+                    exit(1);
+                }
+
+                lexTokens.push(token(line,startingColumn, placeholder));
+            }
+            // if not space, valid operator, or valid number, print error:
+            else {
+                std::cout << "Syntax error on line " << line << " column " << column << ".\n";
+                exit(1);
+            }
         }
-    }
+        // if input is space, add column:
+        column++;
+    } 
+    createEnd(lexTokens, line, column);
+}
+
+
 
```

Comment:
I think we should do 3 newlines between functions of same namespace and 6 lines between different namespace functions